### PR TITLE
[circle-mpqsolver] Introduce PatternSolver

### PR DESCRIPTION
**Pull request by stamalakhov**
This commit introduces PatternSolver
to get optimal mixed precision model.

Its correctness is tested at https://github.com/Samsung/ONE/pull/11560.

Draft: https://github.com/Samsung/ONE/pull/11560
Related: https://github.com/Samsung/ONE/issues/11543

ONE-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

**Review by jinevening — APPROVED** [review]
LGTM

Hmm.. It found that some methods in `MPQSolver` have to be moved to `PatternSolver`.

```cpp
  /**
   * @brief set quantization options
   */
  void set_mpq_options(MPQOptions &options);

  /**
   * @brief fill _frozen with prescribed quantization parameters of resolved nodes
   */
  void resolve_patterns(luci::Module *module);

  /**
   * @brief resolve Q8LayerNormWithQ16Variance pattern
   */
  void resolve_layer_norm_pattern(luci::Module *module);

  /**
   * @brief transform _frozen nodes to Quantizer friendly form
   */
  luci::CircleQuantizer::Options::LayerParams get_frozen_params() const;
```

Those methods are only used in `PatternSolver`.

File: compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  resolve_patterns(module.get());
+
+  auto layer_params = get_frozen_params();
+
+  if (!_quantizer->quantize(module.get(), "uint8", layer_params))
+  {
+    std::cerr << "ERROR: Failed to quantize model" << std::endl;
+    return nullptr;
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp:47]: Q) above throws runtime_error. this line returns nullptr.
what is the decision point for this?

File: compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  MPQOptions options{patterns};
+  set_mpq_options(options);
+}
+
+std::unique_ptr<luci::Module> PatternSolver::run(const std::string &module_path)
+{
+  auto module = read_module(module_path);
+  if (!module)
+  {
+    throw std::runtime_error("Failed to load module");
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp:37]: `module` or `model` ?

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp:37]: Ahh. Sorry . `Model` sounds more user friendly.

**Comment by stamalakhov** [comment]
> Hmm.. I found that some methods in `MPQSolver` have to be moved to `PatternSolver`.

@jinevening 
But they will be useful for `Bisection` solver or any other solvers, which will use frozen parameters.
Should i move them into 'PatternSolver?

File: compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp (hunk 1 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp:47]: I thought that:
throw will notify user that input data is incorrect, while quantize may fail due to incorrect output parameters.
I'll remove `throw` as `Bisection` doesn't throw in that case.

**Comment by jinevening** [comment]
> But they will be useful for Bisection solver or any other solvers, which will use frozen parameters.
Should i move them into 'PatternSolver?

Currently Bisection solver does not use those functions. Do you have a plan to support combination of pattern solver and bisection solver? If so, it would be better to make a new class e.g., `PatternResolver`, which includes those methods, and let `BisectionSolver` and `PatternSolver` inherit `PatternResolver`.

**Comment by stamalakhov** [comment]
> > But they will be useful for Bisection solver or any other solvers, which will use frozen parameters.
> > Should i move them into 'PatternSolver?
> 
> Currently Bisection solver does not use those functions. Do you have a plan to support combination of pattern solver and bisection solver? If so, it would be better to make a new class e.g., `PatternResolver`, which includes those methods, and let `BisectionSolver` and `PatternSolver` inherit `PatternResolver`.
> 
> This can be done in a separate PR.

ok.

**Comment by stamalakhov** [comment]
@jinevening 
done.

File: compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp (hunk 1 above)
Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp:48]: Ahhh. ok.

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp:52]: @seanshpark
done.

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp:52]: For `bisection` it will be included into separate PR.

Comment (seanshpark) [reply compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp:52]: > throw will notify user that input data is incorrect, while quantize may fail due to incorrect output parameters.
I'll remove throw as Bisection doesn't throw in that case.

`std::cerr << "ERROR: Failed to quantize model" << std::endl;` also notifies to the user with same format. I'm still not certain that two may exit in different ways.

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp:52]: ok. I'll change it to throw also.

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp]: done

File: compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  MPQOptions options{patterns};
+  set_mpq_options(options);
+}
+
+std::unique_ptr<luci::Module> PatternSolver::run(const std::string &module_path)
+{
+  auto module = read_module(module_path);
+  if (!module)
+  {
+    throw std::runtime_error("Failed to load model");
```
Comment (seanshpark) [inline compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp:42]: in `MPQSolver::read_module()`,
```cxx
  if (module.get() == nullptr)
  {
    std::cerr << "ERROR: Failed to load " << path << std::endl;
    return nullptr;
  }
```
which seems to show two `Failed to load` message.
IMHO, it maybe difficult to solve this issue in this PR so please fire a new issue for this problem, and go as-is in this PR.

Comment (stamalakhov) [reply compiler/circle-mpqsolver/src/pattern/PatternSolver.cpp:42]: Ahh. Ok.

**Review by jinevening — APPROVED** [review]
LGTM Thank you

**Pull request merged**